examples/nftvm/storage/nft_storage.go

```diff
@@ -0,0 +1,273 @@
+	return name, symbol, metadata, numOfInstances, owner, nil
+}
+
+// This function takes care of the following
+// 1. Creating a NFT Instance by assigning it the next available instance number
+// 2. Updating the parent collection state (i.e. by incrementing numOfInstances)
+func CreateNFTInstance(
+	ctx context.Context,
+	mu state.Mutable,
+	collectionAddress codec.Address,
+	owner codec.Address,
+	metadata []byte,
+) (instanceNum uint32, err error) {
+	// Get parent collection state
+	name, symbol, collectionMetadata, numOfInstances, collectionOwner, err := GetNFTCollectionNoController(ctx, mu, collectionAddress)
+	if err != nil {
+		return 0, err
+	}
+
+	// Write instance to state
+	if err = SetNFTInstance(ctx, mu, collectionAddress, numOfInstances, owner, metadata, false); err != nil {
+		return 0, err
+	}
+
+	// Update parent collection in state
+	if err = SetNFTCollection(ctx, mu, collectionAddress, name, symbol, collectionMetadata, numOfInstances+1, collectionOwner); err != nil {
+		return 0, nil
+	}
+
+	return numOfInstances, nil
+}
+
+// Writes NFT instance to state
+func SetNFTInstance(
+	ctx context.Context,
+	mu state.Mutable,
+	collectionAddress codec.Address,
+	instanceNum uint32,
+	owner codec.Address,
+	metadata []byte,
+	isListedOnMarketplace bool,
+) error {
+	instanceStateKey := InstanceStateKey(collectionAddress, instanceNum)
+
+	metadataLen := len(metadata)
+	v := make([]byte, codec.AddressLen+consts.Uint16Len+metadataLen+consts.Uint16Len)
+
+	// Insert instance owner
+	copy(v, owner[:])
+
+	// Insert metadata
+	binary.BigEndian.PutUint16(v[codec.AddressLen:], uint16(metadataLen))
+	copy(v[codec.AddressLen+consts.Uint16Len:], metadata)
+
+	// Insert marketplace indicator
+	if isListedOnMarketplace {
+		binary.BigEndian.PutUint16(v[codec.AddressLen+consts.Uint16Len+metadataLen:], 1)
+	} else {
+		binary.BigEndian.PutUint16(v[codec.AddressLen+consts.Uint16Len+metadataLen:], 0)
+	}
+
+	return mu.Insert(ctx, instanceStateKey, v)
+}
+
+func GetNFTInstance(
+	ctx context.Context,
+	f ReadState,
+	collectionAddress codec.Address,
+	instanceNum uint32,
+) (owner codec.Address, metadata []byte, isListedOnMarketplace bool, err error) {
+	k := InstanceStateKey(collectionAddress, instanceNum)
+	values, errs := f(ctx, [][]byte{k})
+	return innerGetNFTInstance(values[0], errs[0])
+}
+
+func GetNFTInstanceNoController(
+	ctx context.Context,
+	mu state.Mutable,
+	collectionAddress codec.Address,
+	instanceNum uint32,
+) (owner codec.Address, metadata []byte, isListedOnMarketplace bool, err error) {
+	value, err := mu.GetValue(ctx, InstanceStateKey(collectionAddress, instanceNum))
+	if err != nil {
+		return codec.EmptyAddress, []byte{}, false, err
+	}
+	return innerGetNFTInstance(value, err)
+}
+
+func innerGetNFTInstance(
+	v []byte,
+	err error,
+) (owner codec.Address, metadata []byte, isListedOnMarketplace bool, e error) {
+	if err != nil {
+		return codec.EmptyAddress, []byte{}, false, err
+	}
```

Comment:
We should check the error outside of this function rather than passing it in to do the error check here.

Reply:
we should leave to good design and re-usable code to reduce the amount of code it takes to implement a VM rather than try to eliminate every line possible